src/UI/Game.jsx

```diff
@@ -0,0 +1,458 @@
+      this.props.fetchUser(res.data.token);
+      this.gameOver();
+    }).catch((err) => this.setState((state) => state.modal.data.error = err.response.data.message));
+    event.preventDefault();
+  }
+
+  modal(active, type, data) {
+    keyPressAllow = false;
+    const {icon, title, description, inputId, firstButton, secondButton, submit, error, exit} = data
+
+    this.setState({
+      modal: {
+        active,
+        data: {
+          type,
+          exit: exit || false,
+          icon: icon || null,
+          title: title || null,
+          description: description || null,
+          inputId: inputId || 0,
+          error: error || null,
+          firstButton: {
+            text: firstButton.text || "OK",
+            handler: firstButton.handler || null
+          },
+          secondButton: {
+            text: secondButton.text || null,
+            handler: secondButton.handler || null
+          },
+          submit: submit || ""
+        }
+      }
+    })
+  }
+
+  componentDidMount() {
+    document.addEventListener('keypress', this.handleKeyPress)
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('keypress', this.handleKeyPress)
+  }
+
+  handleKeyPress = async event => {
+    const tempCells = this.props.cells;
+
+    if (event.code === "KeyF") {
+      this.props.changeTheme(!this.props.nightmode)
+      this.saveAllData({...this.state})
+    }
+
+    if (event.code === "KeyR") {
+      this.props.newField(this.props.size);
+      this.saveAllData({...this.state});
+    }
+
+    if (['KeyA', 'KeyS', 'KeyD', 'KeyW'].includes(event.code) && keyPressAllow) {
+
+      this.setState({additionalScore: null})
+
+      keyPressAllow = false;
+
+      const mapKeyCodeToDirection = {
+        KeyA: directions.LEFT,
+        KeyS: directions.DOWN,
+        KeyD: directions.RIGHT,
+        KeyW: directions.UP,
+      }
+
+      this.props.updateField(moveCells(this.props.cells, mapKeyCodeToDirection[event.code], this.props.size))
+
+      const tempScore = this.props.score;
+
+      this.props.cells.filter(cell => cell.state !== "DYING").forEach((cell) => {
+        if (cell.state === "INCREASE") {
+          this.props.addScore(cell.value*2);
+        }
+      });
+
+      if (this.props.score !== tempScore) {
+        this.setState({additionalScore: this.props.score-tempScore});
+      }
+
+      if(isChanged(tempCells, this.props.cells)) {
+        await delay(100);
+
+        if (this.props.sound) {
+          this.moveSound.play();
+        }
+
+        const cells = removeAndIncreaseCells(this.props.cells);
+
+        this.props.updateField(populateField(cells, this.props.size));
+        this.setState(state => ({
+          ...state,
+          bestScore: this.props.score > this.state.bestScore ? this.props.score : this.state.bestScore
+        }))
+        if(this.props.cells && this.props.cells.some(cell => cell.value === 2048) && !this.props.cells.some(cell => cell.value === 4096)) {
```

Comment:
Use constants for these values (2048, 4096, etc.). Otherwise it seems like magic values from nowhere :) And it would be easier to manage them.